File: archive/seema-node/src/routes/intake.ts
import { Router } from 'express';
import { z } from 'zod';
import prisma from '../lib/prisma.js';
import { authenticate, requireRole } from '../middleware/auth.js';
import { getTenantFilter } from '../middleware/tenant.js';
import { logAudit } from '../middleware/auditLogger.js';

const router = Router();

// Serialize Prisma's camelCase to the snake_case shape the frontend expects.
function serializeIntake(i: any) {
  return {
    id: i.id,
    client_name: i.clientName,
    client_email: i.clientEmail,
    client_phone: i.clientPhone,
    client_type: i.clientType,
    company_name: i.companyName,
    practice_area: i.practiceArea,
    risk_level: i.riskLevel,
    risk_score: i.riskScore,
    assigned_to: i.assignedTo,
    source_of_funds: i.sourceOfFunds,
    pep_screening: i.pepScreening,
    sanctions_check: i.sanctionsCheck,
    cdd_status: i.cddStatus,
    conflict_check_status: i.conflictCheckStatus,
    conflict_check_details: i.conflictCheckDetails,
    client_care_letter_sent: i.clientCareLetterSent,
    status: i.status,
    created_at: i.createdAt,
    updated_at: i.updatedAt,
  };
}

// GET /compliance/intake
router.get('/compliance/intake', authenticate, async (req, res, next) => {
  try {
    const { firmId } = getTenantFilter(req);
    const intakes = await prisma.clientIntake.findMany({ where: { firmId } });
    res.json(intakes.map(serializeIntake));
  } catch (err) {
    next(err);
  }
});

// POST /compliance/intake
//
// Accepts either snake_case (the frontend sends this) or camelCase keys —
// the backend was previously requiring camelCase that the UI never sent,
// causing every create to 400. We accept both so existing API clients +
// any new ones don't break.
const createIntakeSchema = z.object({
  // snake_case (frontend)
  client_name: z.string().optional(),
  client_email: z.string().optional(),
  client_phone: z.string().optional(),
  client_type: z.string().optional(),
  company_name: z.string().optional(),
  matter_type: z.string().optional(),
  practice_area: z.string().optional(),
  assigned_fee_earner: z.string().optional(),
  client_reference: z.string().optional(),
  // camelCase (alternative)
  clientName: z.string().optional(),
  clientEmail: z.string().optional(),
  clientPhone: z.string().optional(),
  clientType: z.string().optional(),
  companyName: z.string().optional(),
  practiceArea: z.string().optional(),
  assignedTo: z.string().optional(),
  sourceOfFunds: z.string().optional(),
}).refine(
  (d) => Boolean(d.client_name || d.clientName),
  { message: 'client_name (or clientName) is required' },
);

// Postgres text/varchar columns reject literal NUL (0x00) bytes — strip them
// from any string input before saving. Browsers occasionally smuggle them in
// via autofill or paste from rich-text sources.
const clean = (v: string | null | undefined): string | null => {
  if (v == null) return null;
  const s = String(v).replace(/\u0000/g, '').trim();
  return s.length ? s : null;
};

router.post('/compliance/intake', authenticate, async (req, res, next) => {
  try {
    const { firmId } = getTenantFilter(req);
    const data = createIntakeSchema.parse(req.body);

    const intake = await prisma.clientIntake.create({
      data: {
        firmId,
        clientName: clean(data.client_name ?? data.clientName) as string,
        clientEmail: clean(data.client_email ?? data.clientEmail),
        clientPhone: clean(data.client_phone ?? data.clientPhone),
        clientType: clean(data.client_type ?? data.clientType),
        companyName: clean(data.company_name ?? data.companyName),
        // matter_type maps to practiceArea — we treat them as the same field
        practiceArea: clean(data.practice_area ?? data.practiceArea ?? data.matter_type),
        assignedTo: clean(data.assigned_fee_earner ?? data.assignedTo),
        sourceOfFunds: clean(data.client_reference ?? data.sourceOfFunds),
        status: 'pending',
      },
    });

    res.status(201).json(serializeIntake(intake));
  } catch (err) {
    next(err);
  }
});

// PUT /compliance/intake/:intakeId
//
// Frontend sends snake_case (client_name, client_email, etc.); older callers
// sent camelCase. Accept both — every field optional for partial updates.
const updateIntakeSchema = z.object({
  // snake_case (frontend)
  client_name: z.string().optional(),
  client_email: z.string().optional(),
  client_phone: z.string().optional(),
  client_type: z.string().optional(),
  company_name: z.string().optional(),
  matter_type: z.string().optional(),
  practice_area: z.string().optional(),
  assigned_fee_earner: z.string().optional(),
  client_reference: z.string().optional(),
  risk_level: z.string().optional(),
  // camelCase (legacy)
  clientName: z.string().optional(),
  clientEmail: z.string().optional(),
  clientPhone: z.string().optional(),
  clientType: z.string().optional(),
  companyName: z.string().optional(),
  practiceArea: z.string().optional(),
  matterType: z.string().optional(),
  assignedTo: z.string().optional(),
  riskLevel: z.string().optional(),
  sourceOfFunds: z.string().optional(),
  email: z.string().email().optional(),
  phone: z.string().optional(),
  notes: z.string().optional(),
  status: z.string().optional(),
});

router.put('/compliance/intake/:intakeId', authenticate, async (req, res, next) => {
  try {
    const { firmId } = getTenantFilter(req);
    const data = updateIntakeSchema.parse(req.body);

    const updateData: Record<string, unknown> = {};
    const clientName = data.client_name ?? data.clientName;
    if (clientName !== undefined) updateData.clientName = clean(clientName);
    const clientEmail = data.client_email ?? data.clientEmail ?? data.email;
    if (clientEmail !== undefined) updateData.clientEmail = clean(clientEmail);
    const clientPhone = data.client_phone ?? data.clientPhone ?? data.phone;
    if (clientPhone !== undefined) updateData.clientPhone = clean(clientPhone);
    const clientType = data.client_type ?? data.clientType;
    if (clientType !== undefined) updateData.clientType = clean(clientType);
    const companyName = data.company_name ?? data.companyName;
    if (companyName !== undefined) updateData.companyName = clean(companyName);
    const practiceArea = data.practice_area ?? data.practiceArea ?? data.matter_type ?? data.matterType;
    if (practiceArea !== undefined) updateData.practiceArea = clean(practiceArea);
    const assignedTo = data.assigned_fee_earner ?? data.assignedTo;
    if (assignedTo !== undefined) updateData.assignedTo = clean(assignedTo);
    const sourceOfFunds = data.client_reference ?? data.sourceOfFunds;
    if (sourceOfFunds !== undefined) updateData.sourceOfFunds = clean(sourceOfFunds);
    const riskLevel = data.risk_level ?? data.riskLevel;
    if (riskLevel !== undefined) updateData.riskLevel = clean(riskLevel);
    if (data.status !== undefined) updateData.status = data.status;

    const result = await prisma.clientIntake.updateMany({
      where: { id: (req.params.intakeId as string), firmId },
      data: updateData,
    });

    if (result.count === 0) {
      res.status(404).json({ error: true, message: 'Intake not found' });
      return;
    }

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'intake_updated',
      entityType: 'client_intake',
      entityId: (req.params.intakeId as string),
    });

    const updated = await prisma.clientIntake.findFirst({ where: { id: (req.params.intakeId as string), firmId } });
    res.json(updated ? serializeIntake(updated) : null);
  } catch (err) {
    next(err);
  }
});

// POST /compliance/intake/:intakeId/approve
router.post('/compliance/intake/:intakeId/approve', authenticate, requireRole('admin'), async (req, res, next) => {
  try {
    const { firmId } = getTenantFilter(req);

    const result = await prisma.clientIntake.updateMany({
      where: { id: (req.params.intakeId as string), firmId },
      data: { status: 'approved' },
    });

    if (result.count === 0) {
      res.status(404).json({ error: true, message: 'Intake not found' });
      return;
    }

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'intake_approved',
      entityType: 'client_intake',
      entityId: (req.params.intakeId as string),
    });

    const updated = await prisma.clientIntake.findFirst({ where: { id: (req.params.intakeId as string), firmId } });
    res.json(updated ? serializeIntake(updated) : null);
  } catch (err) {
    next(err);
  }
});

// POST /compliance/intake/:intakeId/reject
router.post('/compliance/intake/:intakeId/reject', authenticate, requireRole('admin'), async (req, res, next) => {
  try {
    const { firmId } = getTenantFilter(req);

    const result = await prisma.clientIntake.updateMany({
      where: { id: (req.params.intakeId as string), firmId },
      data: { status: 'rejected' },
    });

    if (result.count === 0) {
      res.status(404).json({ error: true, message: 'Intake not found' });
      return;
    }

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'intake_rejected',
      entityType: 'client_intake',
      entityId: (req.params.intakeId as string),
    });

    const updated = await prisma.clientIntake.findFirst({ where: { id: (req.params.intakeId as string), firmId } });
    res.json(updated ? serializeIntake(updated) : null);
  } catch (err) {
    next(err);
  }
});

export default router;

File: backend/routers/intake.py
"""Intake router — client intake and CDD management."""
import uuid
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from middleware.tenant_rls import tenant_db_from_jwt, bypass_db
from middleware.auth import get_current_user, CurrentUser
from services.audit_logger import log_audit
from models.intake import ClientIntake

router = APIRouter()


class _AssessIntakeBody(BaseModel):
    risk_assessment: str
    notes: str | None = None


@router.post("/compliance/intake/{intake_id}/assess")
async def assess_intake(
    intake_id: str,
    body: _AssessIntakeBody,
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Record a CDD risk assessment on an intake (was 404)."""
    res = await db.execute(
        select(ClientIntake).where(
            ClientIntake.id == intake_id,
            ClientIntake.firm_id == current_user.firm_id,
        )
    )
    intake = res.scalar_one_or_none()
    if not intake:
        raise HTTPException(status_code=404, detail="Intake not found")
    intake.risk_level = body.risk_assessment
    intake.risk_score = {"low": 20, "medium": 50, "high": 80}.get(body.risk_assessment, 50)
    intake.status = "reviewed"
    intake.cdd_status = "assessed"
    if body.notes:
        intake.conflict_check_details = body.notes
    await db.flush()
    await log_audit(
        db=db, firm_id=current_user.firm_id, action="assessed",
        entity_type="client_intake", entity_id=intake.id, user_id=current_user.user_id,
        details=f"Risk assessed as {body.risk_assessment}",
    )
    return {"id": intake.id, "risk_level": intake.risk_level,
            "risk_score": intake.risk_score, "status": intake.status}

class CreateIntakeRequest(BaseModel):
    client_name: str
    client_email: str | None = None
    practice_area: str | None = None
    risk_level: str = "medium"
    source_of_funds: str | None = None

class UpdateIntakeRequest(BaseModel):
    client_name: str | None = None
    client_email: str | None = None
    practice_area: str | None = None
    risk_level: str | None = None
    source_of_funds: str | None = None
    assigned_to: str | None = None
    conflict_check_status: str | None = None
    cdd_status: str | None = None

@router.get("/compliance/intake")
async def get_intakes(
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """List intakes for firm."""
    result = await db.execute(
        select(ClientIntake).where(
            ClientIntake.firm_id == user.firm_id
        ).order_by(ClientIntake.created_at.desc())
    )
    intakes = result.scalars().all()

    return [
        {
            "id": i.id,
            "client_name": i.client_name,
            "client_email": i.client_email,
            "practice_area": i.practice_area,
            "status": i.status,
            "conflict_check_status": i.conflict_check_status,
            "client_care_letter_sent": i.client_care_letter_sent,
            "risk_level": i.risk_level,
            "risk_score": float(i.risk_score) if i.risk_score else 0,
            "assigned_to": i.assigned_to,
            "source_of_funds": i.source_of_funds,
            "pep_screening": i.pep_screening,
            "sanctions_check": i.sanctions_check,
            "cdd_status": i.cdd_status,
            "created_at": str(i.created_at),
            "updated_at": str(i.updated_at),
        }
        for i in intakes
    ]

@router.post("/compliance/intake")
async def create_intake(
    req: CreateIntakeRequest,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Create new intake."""
    intake = ClientIntake(
        id=str(uuid.uuid4()),
        firm_id=user.firm_id,
        client_name=req.client_name,
        client_email=req.client_email,
        practice_area=req.practice_area,
        risk_level=req.risk_level,
        source_of_funds=req.source_of_funds,
        status="pending",
    )
    db.add(intake)
    await db.flush()

    await log_audit(
        db=db,
        firm_id=user.firm_id,
        action="created",
        entity_type="client_intake",
        entity_id=intake.id,
        user_id=user.user_id,
        details=f"Client intake created: {req.client_name}",
    )

    return {
        "id": intake.id,
        "client_name": intake.client_name,
        "status": intake.status,
        "created_at": str(intake.created_at),
    }

@router.put("/compliance/intake/{intake_id}")
async def update_intake(
    intake_id: str,
    req: UpdateIntakeRequest,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Update intake."""
    result = await db.execute(
        select(ClientIntake).where(
            ClientIntake.id == intake_id,
            ClientIntake.firm_id == user.firm_id,
        )
    )
    intake = result.scalar_one_or_none()
    if not intake:
        raise HTTPException(status_code=404, detail="Intake not found")

    if req.client_name is not None:
        intake.client_name = req.client_name
    if req.client_email is not None:
        intake.client_email = req.client_email
    if req.practice_area is not None:
        intake.practice_area = req.practice_area
    if req.risk_level is not None:
        intake.risk_level = req.risk_level
    if req.source_of_funds is not None:
        intake.source_of_funds = req.source_of_funds
    if req.assigned_to is not None:
        intake.assigned_to = req.assigned_to
    if req.conflict_check_status is not None:
        intake.conflict_check_status = req.conflict_check_status
    if req.cdd_status is not None:
        intake.cdd_status = req.cdd_status

    await db.flush()

    await log_audit(
        db=db,
        firm_id=user.firm_id,
        action="updated",
        entity_type="client_intake",
        entity_id=intake_id,
        user_id=user.user_id,
        details=f"Intake updated: {intake.client_name}",
    )

    return {
        "id": intake.id,
        "client_name": intake.client_name,
        "status": intake.status,
        "updated_at": str(intake.updated_at),
    }

@router.post("/compliance/intake/{intake_id}/approve")
async def approve_intake(
    intake_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Approve intake."""
    result = await db.execute(
        select(ClientIntake).where(
            ClientIntake.id == intake_id,
            ClientIntake.firm_id == user.firm_id,
        )
    )
    intake = result.scalar_one_or_none()
    if not intake:
        raise HTTPException(status_code=404, detail="Intake not found")

    intake.status = "approved"
    await db.flush()

    await log_audit(
        db=db,
        firm_id=user.firm_id,
        action="approved",
        entity_type="client_intake",
        entity_id=intake_id,
        user_id=user.user_id,
        details=f"Intake approved: {intake.client_name}",
    )

    return {"id": intake.id, "status": intake.status}

@router.post("/compliance/intake/{intake_id}/reject")
async def reject_intake(
    intake_id: str,
    user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Reject intake."""
    result = await db.execute(
        select(ClientIntake).where(
            ClientIntake.id == intake_id,
            ClientIntake.firm_id == user.firm_id,
        )
    )
    intake = result.scalar_one_or_none()
    if not intake:
        raise HTTPException(status_code=404, detail="Intake not found")

    intake.status = "rejected"
    await db.flush()

    await log_audit(
        db=db,
        firm_id=user.firm_id,
        action="rejected",
        entity_type="client_intake",
        entity_id=intake_id,
        user_id=user.user_id,
        details=f"Intake rejected: {intake.client_name}",
    )

    return {"id": intake.id, "status": intake.status}
